### PR TITLE
db4: fix configure with llvm

### LINKED ISSUE #315245: Build failure: pkgsLLVM.db
### Steps To Reproduce

1. build `pkgsLLVM.db`

### Build log

```
checking size of char *... 8
checking for u_char... yes
checking for u_short... yes
checking for u_int... yes
checking for u_long... yes
checking for u_int8_t... yes
checking for u_int16_t... yes
checking for int16_t... yes
checking for u_int32_t... yes
checking for int32_t... yes
checking for u_int64_t... yes
checking for int64_t... yes
checking for FILE *... yes
checking for off_t... yes
checking for pid_t... yes
checking for size_t... yes
checking for time_t... yes
checking size of size_t... 8
checking for ssize_t... yes
checking for uintmax_t... yes
checking for uintptr_t... yes
checking for socklen_t... yes
checking for ANSI C exit success/failure values... yes
checking for getopt optreset variable... no
checking for mutexes... UNIX/fcntl
configure: WARNING: NO SHARED LATCH IMPLEMENTATION FOUND FOR THIS PLATFORM.
configure: error: Unable to find a mutex implementation
```

### Notify maintainers

@reckenrode 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"aarch64-linux"`
 - host os: `Linux 5.15.156, NixOS, 24.05 (Uakari), 24.05.20240418.bc279bb`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/run/current-system/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Description of changes

Can be tested by building `pkgsLLVM.db4` (`nix-build -A pkgsLLVM.db4` / `nix build .#pkgsLLVM.db4`)

Fixes #315245

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/db/db-4.8.nix [hunk 1]
```diff
@@ -9,6 +9,8 @@ import ./generic.nix (args // {
     ./darwin-mutexes-4.8.patch
   ];
 
+  drvArgs.configureFlags = lib.optional (targetPlatform.useLLVM or false) "--with-mutex=POSIX/pthreads";
```
Comment (drupol) [inline pkgs/development/libraries/db/db-4.8.nix:12]:
```suggestion
  drvArgs.configureFlags = lib.optionals (targetPlatform.useLLVM or false) [ "--with-mutex=POSIX/pthreads" ];
```

File: pkgs/development/libraries/db/generic.nix [hunk 2]
```diff
@@ -67,7 +67,8 @@ stdenv.mkDerivation (rec {
       (if compat185 then "--enable-compat185" else "--disable-compat185")
     ]
     ++ lib.optional dbmSupport "--enable-dbm"
-    ++ lib.optional stdenv.isFreeBSD "--with-pic";
+    ++ lib.optional stdenv.isFreeBSD "--with-pic"
+    ++ (drvArgs.configureFlags or []);
```
Comment (drupol) [inline pkgs/development/libraries/db/generic.nix:71]:
```suggestion
    ++ lib.optionals stdenv.isFreeBSD [ "--with-pic" ]
    ++ (drvArgs.configureFlags or []);
```

**Review by drupol — COMMENTED** [review]
**suggestion**: I would use `lib.optionals` here.

Using https://conventionalcomments.org/ for the first time on `nixpkgs`!

**Comment by RossComputerGuy** [comment]
@drupol Why `optionals` over `optional` with single item lists?

**Comment by paparodeo** [comment]
the issue here is the configure test is using exit but not including the correct header file, thus causing the test to fail.

```
configure:21863: clang -o conftest -O3   -D_GNU_SOURCE -D_REENTRANT  conftest.c  >&5
conftest.c:52:2: error: call to undeclared library function 'exit' with type 'void (int) __attribute__((noreturn))'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   52 |         exit (
      |         ^
conftest.c:52:2: note: include the header <stdlib.h> or explicitly provide a declaration for 'exit'
```

there is already a patch to dist/aclocal/mutex.m4 to fix a similar issue -- no return time for function definition -- and can be modified to just change the exit -> return.

eg:

```diff
diff --git a/pkgs/development/libraries/db/clang-4.8.patch b/pkgs/development/libraries/db/clang-4.8.patch
index 0aff87682037..3f614363af80 100644
--- a/pkgs/development/libraries/db/clang-4.8.patch
+++ b/pkgs/development/libraries/db/clang-4.8.patch
@@ -153,7 +153,7 @@ diff -ur a/dist/aclocal/clock.m4 b/dist/aclocal/clock.m4
 diff -ur a/dist/aclocal/mutex.m4 b/dist/aclocal/mutex.m4
 --- a/dist/aclocal/mutex.m4	1969-12-31 19:00:01.000000000 -0500
 +++ b/dist/aclocal/mutex.m4	2023-06-05 19:14:47.214158196 -0400
-@@ -4,6 +4,7 @@
+@@ -4,12 +4,13 @@
  AC_DEFUN(AM_PTHREADS_SHARED, [
  AC_TRY_RUN([
  #include <pthread.h>
@@ -161,7 +161,23 @@ diff -ur a/dist/aclocal/mutex.m4 b/dist/aclocal/mutex.m4
  main() {
  	pthread_cond_t cond;
  	pthread_mutex_t mutex;
-@@ -46,6 +47,7 @@
+ 	pthread_condattr_t condattr;
+ 	pthread_mutexattr_t mutexattr;
+-	exit (
++	return (
+ 	pthread_condattr_init(&condattr) ||
+ 	pthread_condattr_setpshared(&condattr, PTHREAD_PROCESS_SHARED) ||
+ 	pthread_mutexattr_init(&mutexattr) ||
+@@ -29,7 +30,7 @@ AC_TRY_LINK([
+ 	pthread_mutex_t mutex;
+ 	pthread_condattr_t condattr;
+ 	pthread_mutexattr_t mutexattr;
+-	exit (
++	return (
+ 	pthread_condattr_init(&condattr) ||
+ 	pthread_condattr_setpshared(&condattr, PTHREAD_PROCESS_SHARED) ||
+ 	pthread_mutexattr_init(&mutexattr) ||
+@@ -46,12 +47,13 @@ AC_TRY_LINK([
  AC_DEFUN(AM_PTHREADS_PRIVATE, [
  AC_TRY_RUN([
  #include <pthread.h>
@@ -169,6 +185,22 @@ diff -ur a/dist/aclocal/mutex.m4 b/dist/aclocal/mutex.m4
  main() {
  	pthread_cond_t cond;
  	pthread_mutex_t mutex;
+ 	pthread_condattr_t condattr;
+ 	pthread_mutexattr_t mutexattr;
+-	exit (
++	return (
+ 	pthread_condattr_init(&condattr) ||
+ 	pthread_mutexattr_init(&mutexattr) ||
+ 	pthread_cond_init(&cond, &condattr) ||
+@@ -69,7 +71,7 @@ AC_TRY_LINK([
+ 	pthread_mutex_t mutex;
+ 	pthread_condattr_t condattr;
+ 	pthread_mutexattr_t mutexattr;
+-	exit (
++	return (
+ 	pthread_condattr_init(&condattr) ||
+ 	pthread_mutexattr_init(&mutexattr) ||
+ 	pthread_cond_init(&cond, &condattr) ||
 diff -ur a/dist/aclocal/sequence.m4 b/dist/aclocal/sequence.m4
 --- a/dist/aclocal/sequence.m4	1969-12-31 19:00:01.000000000 -0500
 +++ b/dist/aclocal/sequence.m4	2023-06-05 19:14:02.007869956 -0400
```

**Pull request merged**